components/esp_dns/README.md

```diff
@@ -0,0 +1,189 @@
+
+## Table of Contents
+- [Features](#features)
+- [How It Works](#how-it-works)
+- [Certificate Options](#certificate-options)
+- [How to Use](#how-to-use)
+- [Configuration](#configuration)
+- [Common DNS Servers](#common-dns-servers)
+- [Requirements](#requirements)
+- [Limitations](#limitations)
+- [Performance Considerations](#performance-considerations)
+- [Troubleshooting](#troubleshooting)
+
+## Features
+
+- **Multiple Protocol Support** Choose from various DNS protocols:
+  - Standard UDP DNS (Port 53)
+  - TCP DNS (Port 53)
+  - DNS over TLS (DoT) (Port 853)
+  - DNS over HTTPS (DoH) (Port 443)
+
+- **Secure DNS Resolution**: Supports encrypted DNS queries using TLS and HTTPS to protect privacy and prevent DNS spoofing.
+
+- **Flexible Configuration**: Easily configure DNS servers, ports, timeouts, and protocol-specific options.
+
+- **LWIP Integration**: Seamlessly integrates with the ESP-IDF networking stack through LWIP hooks.
+
+- **Standard getaddrinfo() Interface**: Use the standard `getaddrinfo()` function to resolve domain names.
+
+## How It Works
+This component utilizes the `CONFIG_LWIP_HOOK_NETCONN_EXT_RESOLVE_CUSTOM` hook to override the core DNS functionality of LWIP and implement custom DNS over HTTPS resolution. To enable this, ensure that the configuration option `Component config → LWIP → Hooks → Netconn external resolve Hook` is set to `Custom implementation`.
+
+Once you add this component to your project, it will replace the default LWIP DNS resolution automatically.
+
+**⚠️ Warning:** This component cannot work alongside other components that use the CONFIG_LWIP_HOOK_NETCONN_EXT_RESOLVE_CUSTOM hook, such as the OpenThread component.
+
+## Certificate Options
+
+When using secure DNS protocols (DoT and DoH), you have two certificate options:
+
+1. **Certificate Bundle**: Use ESP-IDF's certificate bundle for validating connections to popular DNS providers.
+2. **Custom Certificate**: Provide your own certificate in PEM format for custom DNS servers.
+
+
+## How to Use
+
+### 1. Add Component to Your Project
+
+Add the ESP DNS component to your project's `components` directory or include it as a dependency in your project's `idf_component.yml` file.
+
+### 2. Enable custom DNS resolution
+
+To enable custom DNS resolution, configure the `CONFIG_LWIP_HOOK_NETCONN_EXT_RESOLVE_CUSTOM` setting either through menuconfig or by adding `CONFIG_LWIP_HOOK_NETCONN_EXT_RESOLVE_CUSTOM=y` to your `sdkconfig.defaults` file to pre-set the configuration during the build process.
+
+### 3. Configure DNS Settings
+
+Initialize the DNS component with your preferred configuration:
+```C
+#include "esp_dns.h"
+
+// Configure DNS over HTTPS
+esp_dns_config_t dns_config = {
+    .protocol = DNS_PROTOCOL_DOH,
+    .dns_server = "dns.google",
+    .port = 443, // Optional, default is 443
+    .timeout_ms = 5000, // Optional, default is 5000
+    .tls_config = {
+        .crt_bundle_attach = esp_crt_bundle_attach, // Optional, default is NULL
```

Comment:
`esp_crt_bundle_attach` is not optional, right, if it defaults to NULL?

Reply:
`.crt_bundle_attach` is optional. Updated the comment for better clarity.